File: NodeJS/MessageServer/src/Messages/messages.ts

enum MsgType {
    Error = 0,
    Private = 1,
    CreateGroup,
    InviteToGroup,
    DeclineRequest,
    AcceptRequest,
    LeaveGroup,
    RemoveGroup

}
class Message {
    senderID: string;
    receiverID: string;
    type: MsgType;
    content: string;
    constructor(senderID: string, receiverID: string) {
        this.senderID = senderID;
        this.receiverID = receiverID;
    }
}

class ErrorMessage extends Message {
    type = MsgType.Error
    constructor(senderID: string, receiverID: string, text: string) {
        super(senderID, receiverID);
        this.content = text;
    }
}


class PrivateMessage extends Message {
    type = MsgType.Private
    constructor(senderID: string, receiverID: string, text: string) {
        super(senderID, receiverID);
        this.content = text;
    }
}
class InviteToGroup extends Message {
    type = MsgType.InviteToGroup;
    constructor(senderID: string, receiverID: string, user: string) {
        super(senderID, receiverID);
        this.content = user;
    }
}
class DeclineRequest extends Message {
    type = MsgType.DeclineRequest;
    constructor(senderID: string, group: string) {
        super(senderID, group);
    }
}
class AcceptRequest extends Message {
    type = MsgType.AcceptRequest;
    constructor(senderID: string, group: string) {
        super(senderID, group);
    }
}
class LeaveGroup extends Message {
    type = MsgType.LeaveGroup;
    constructor(senderID: string, group: string) {
        super(senderID, group);
    }
}
class RemoveGroup extends Message {
    type = MsgType.RemoveGroup;
    constructor(senderID: string, groupname: string) {
        super(senderID, "server");
        this.content = groupname;
    }
}
class CreateGroup extends Message {
    type = MsgType.CreateGroup;
    constructor(senderID: string, groupname: string) {
        super(senderID, "server");
        this.content = groupname;
    }
}



export { MsgType, Message, PrivateMessage, InviteToGroup, DeclineRequest, AcceptRequest, LeaveGroup, RemoveGroup, CreateGroup, ErrorMessage };

File: MessageServer/RxKotlin/MessageServer/src/main/kotlin/at/mh/kotlin/message/server/messages/Messages.kt
package at.mh.kotlin.message.server.messages

import com.google.gson.JsonObject

enum class MessageType {
    Error,
    Private,
    CreateGroup,
    InviteToGroup,
    DeclineRequest,
    AcceptRequest,
    LeaveGroup,
    RemoveGroup

}

abstract class Message(var senderID: String, var receiverID: String, var content: String, var type: MessageType) {
    override fun toString(): String {
        val json = JsonObject()
        json.addProperty("senderID", senderID)
        json.addProperty("receiverID", receiverID)
        json.addProperty("content", content)
        json.addProperty("type", type.ordinal)
        return json.toString()
    }
}

class ErrorMessage(senderID: String, receiverID: String, text: String) :
    Message(senderID, receiverID, text, MessageType.Error)

class PrivateMessage(senderID: String, receiverID: String, content: String) :
    Message(senderID, receiverID, content, MessageType.Private)

class InviteToGroup(senderID: String, receiverID: String, inviteeID: String) :
    Message(senderID, receiverID, inviteeID, MessageType.InviteToGroup)

class DeclineRequest(senderID: String, groupID: String) : Message(senderID, groupID, "", MessageType.DeclineRequest)

class AcceptRequest(senderID: String, groupID: String) : Message(senderID, groupID, "", MessageType.AcceptRequest)

class LeaveGroup(senderID: String, groupID: String) : Message(senderID, groupID, "", MessageType.LeaveGroup)

class RemoveGroup(senderID: String, groupID: String) : Message(senderID, "server", groupID, MessageType.RemoveGroup)

class CreateGroup(senderID: String, groupID: String) :
    Message(senderID, "server", groupID, MessageType.CreateGroup)


